src/Stratis.Features.FederatedPeg/SourceChain/MaturedBlocksProvider.cs

```diff
@@ -60,30 +67,55 @@ public async Task<Result<List<MaturedBlockDepositsModel>>> GetMaturedDepositsAsy
             int maxTimeCollectionCanTakeMs = RestApiClientBase.TimeoutMs / 2;
             var cancellation = new CancellationTokenSource(maxTimeCollectionCanTakeMs);
 
-            for (int i = blockHeight; (i <= matureTipHeight) && (i < blockHeight + maxBlocks); i++)
+            int maxBlockHeight = Math.Min(matureTipHeight, blockHeight + maxBlocks - 1);
+
+            var headers = new List<ChainedHeader>();
+            ChainedHeader header = consensusTip.GetAncestor(maxBlockHeight);
+            for (int i = maxBlockHeight; i >= blockHeight; i--)
             {
-                ChainedHeader currentHeader = consensusTip.GetAncestor(i);
+                headers.Add(header);
+                header = header.Previous;
+            }
 
-                ChainedHeaderBlock block = this.consensusManager.GetBlockData(currentHeader.HashBlock);
+            headers.Reverse();
 
-                if (block?.Block?.Transactions == null)
-                {
-                    // Report unexpected results from consenus manager.
-                    this.logger.LogWarning("Stop matured blocks collection due to consensus manager integrity failure. Send what we've collected.");
-                    break;
-                }
+            int numDeposits = 0;
 
-                MaturedBlockDepositsModel maturedBlockDeposits = this.depositExtractor.ExtractBlockDeposits(block);
+            for (int ndx = 0; ndx < headers.Count; ndx += 100)
+            {
+                List<ChainedHeader> currentHeaders = headers.GetRange(ndx, Math.Min(100, headers.Count - ndx));
 
-                if (maturedBlockDeposits == null)
-                    throw new InvalidOperationException($"Unable to get deposits for block at height {currentHeader.Height}");
+                List<uint256> hashes = currentHeaders.Select(h => h.HashBlock).ToList();
 
-                maturedBlocks.Add(maturedBlockDeposits);
+                ChainedHeaderBlock[] blocks = this.consensusManager.GetBlockData(hashes);
 
-                if (cancellation.IsCancellationRequested && maturedBlocks.Count > 0)
+                foreach (ChainedHeaderBlock chainedHeaderBlock in blocks)
                 {
-                    this.logger.LogDebug("Stop matured blocks collection because it's taking too long. Send what we've collected.");
-                    break;
+                    if (chainedHeaderBlock?.Block?.Transactions == null)
+                    {
+                        this.logger.LogDebug("Unexpected null data. Send what we've collected.");
+
+                        return Result<List<MaturedBlockDepositsModel>>.Ok(maturedBlocks);
+                    }
+
+                    MaturedBlockDepositsModel maturedBlockDeposits = this.depositExtractor.ExtractBlockDeposits(chainedHeaderBlock);
```

Comment:
If you look in `ExtractBlockDeposits` you will see that we return if `chainedHeaderBlock` is null. But above we check for it. So I would perhaps remove the null check in `ExtractBlockDeposits`

Reply:
Replace with  Guard rather in `ExtractBlockDeposits`

Reply:
Done.